crates/paimon/src/spec/types.rs

```diff
@@ -0,0 +1,1463 @@
+                        .into_iter()
+                        .map(|p| p.to_string())
+                        .collect::<Vec<_>>()
+                        .join(", ")
+                )
+            )
+        } else {
+            format(
+                format,
+                params
+                    .into_iter()
+                    .map(|p| p.to_string())
+                    .collect::<Vec<_>>()
+                    .join(", "),
+            )
+        }
+    }
+
+    fn to_string(&self) -> String {
+        self.as_sql_string()
+    }
+
+    fn accept<R>(&self, visitor: &dyn DataTypeVisitor<R>) -> R;
+
+    fn collect_field_ids(&self, field_ids: &mut HashSet<i32>) {}
+
+    fn not_null(&self) -> Self {
+        self.copy(false)
+    }
+
+    fn nullable(&self) -> Self {
+        self.copy(true)
+    }
+}
+
+impl DataTypeTrait for DataType {
+    fn copy(&self, is_nullable: bool) -> Self {
+        panic!("Unsupported copy operation.");
+    }
+
+    fn as_sql_string(&self) -> String {
+        panic!("Unsupported as_sql_string operation.");
+    }
+
+    fn accept<R>(&self, visitor: &dyn DataTypeVisitor<R>) -> R {
+        panic!("Unsupported accept operation.");
+    }
+}
+
+impl DataType {
+    pub fn new(is_nullable: bool, root: DataTypeRoot) -> Box<dyn DataType> {
+        DataType {
+            is_nullable,
+            type_root: root,
+        }
+    }
+}
+
+impl Display for DataType {
+    fn fmt(&self, _: &mut Formatter<'_>) -> std::fmt::Result {
+        todo!()
+    }
+}
+
+impl FromStr for DataType {
+    type Err = Error;
+
+    fn from_str(_: &str) -> Result<Self, Self::Err> {
+        todo!()
+    }
+}
+
+/// ArrayType for paimon.
+///
+/// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/ArrayType.java>.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ArrayType {
+    pub element_type: DataType,
+}
+
+impl DataTypeTrait for ArrayType {
+    fn copy(&self, is_nullable: bool) -> Self {
+        ArrayType::new(is_nullable, self.element_type)
+    }
+
+    fn as_sql_string(&self) -> String {
+        format!(Self::FORMAT.replace("%s", &self.get_element_type().as_sql_string()))
+            + if self.data_type.is_nullable {
+                " NULL"
+            } else {
+                ""
+            }
+    }
+
+    fn serialize_json<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("ArrayType", 2)?;
+        state.serialize_field(
+            "type",
+            if self.data_type.is_nullable {
+                "ARRAY"
+            } else {
+                "ARRAY NOT NULL"
+            },
+        )?;
+        state.serialize_field("element", &self.get_element_type())?;
+        state.end()
+    }
+
+    fn equals(&self, other: &Self) -> bool {
```

Comment:
Rust doesn't implement `equals` or `hash` in this way. We need to use `Hash` and `Eq`/`PartialEq`.